lib/docs-renderer/src/renderer.rs

```diff
@@ -0,0 +1,444 @@
+        }
+
+        // The root path always exists.
+        if path == "/" {
+            return true;
+        }
+
+        self.root.pointer(path).is_some()
+    }
+
+    /// Merges the data from `other` into `self`.
+    ///
+    /// Uses a "deep" merge strategy, which will recursively merge both objects together. This
+    /// strategy behaves as follows:
+    ///
+    /// - strings, booleans, integers, numbers, and nulls are "highest priority wins" (`self` has
+    ///   highest priority)
+    /// - arrays are merged together without any deduplication, with the items from `self` appearing
+    ///   first
+    /// - objects have their properties merged together, but if an overlapping property is
+    ///   encountered:
+    ///   - if it has the same type on both sides, the property is merged normally (using the
+    ///     standard merge behavior)
+    ///   - if it does not have the same type on both sides, the property value on the `self` side
+    ///     takes precedence
+    ///
+    /// The only exception to the merge behavior above is if an overlapping object property does not
+    /// have the same type on both sides, but the type on the `self` side is an array. When the type
+    /// is an array, the value on the `other` side is appended to that array, regardless of the
+    /// contents of the array.
+    pub fn merge(&mut self, _other: Self) {
+        todo!()
+    }
+}
+
+impl Default for RenderData {
+    fn default() -> Self {
+        Self {
+            root: Value::Object(Map::new()),
+        }
+    }
+}
+
+pub struct SchemaRenderer<'a, T> {
+    querier: &'a SchemaQuerier,
+    schema: T,
+    data: RenderData,
+}
+
+impl<'a, T> SchemaRenderer<'a, T>
+where
+    T: QueryableSchema,
+{
+    pub fn new(querier: &'a SchemaQuerier, schema: T) -> Self {
+        Self {
+            querier,
+            schema,
+            data: RenderData::default(),
+        }
+    }
+
+    pub fn render(self) -> Result<RenderData, RenderError> {
+        let Self {
+            querier,
+            schema,
+            mut data,
+        } = self;
+
+        // If a schema is hidden, then we intentionally do not want to render it.
+        if schema.has_flag_attribute(constants::DOCS_META_HIDDEN)? {
+            debug!("Schema is marked as hidden. Skipping rendering.");
+
+            return Ok(data);
+        }
+
+        // If a schema is marked as a cycle entrypoint, this means that we would create a cycle (in
+        // the graph sense) if we continued trying to render it.
+        //
+        // This is necessary for some schemas that have a cyclical dependency on each other (like
+        // the `pipelines` transform referring to `Transforms`, which refers to `pipelines`,
+        // and...). We still need to return _something_, so we return some barebones render data --
+        // basically just a description -- so any configuration type that is a cycle entrypoint
+        // should take care to provide a _good_ description.
```

Comment:
Do we still have those, IIRC `pipelines` have been removed.

Reply:
Actually, yeah, after taking a second look, we don't need/use this anymore... so I can remove it here _and_ in the Ruby script, too!

Will fix.

Reply:
Cleaned up in 03262ef97.